src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java

```diff
@@ -374,22 +425,27 @@ private String getUsernameFromProfile(Saml2Profile saml2Profile) {
    * they can configure their IdP.
    */
   public HttpResponse doMetadata(StaplerRequest request, StaplerResponse response) {
-    Saml2Client client = newClient();
+    final SAML2Client client = newClient(request,response);
 
-    return HttpResponses.plainText(client.printClientMetadata());
+    return HttpResponses.plainText(client.getServiceProviderMetadataResolver().getMetadata());
   }
 
-  private Saml2Client newClient() {
+  private SAML2Client newClient(StaplerRequest request, StaplerResponse response) {
     Preconditions.checkNotNull(idpMetadata);
+    final SAML2ClientConfiguration client = new SAML2ClientConfiguration();
+    client.setIdentityProviderMetadataResource(new RewriteableStringResource(idpMetadata));
+    client.setServiceProviderMetadataResource(new RewriteableStringResource());
 
-    Saml2Client client = new Saml2Client();
-    client.setIdpMetadata(idpMetadata);
-    client.setCallbackUrl(getConsumerServiceUrl());
-    client.setDestinationBindingType(SAMLConstants.SAML2_REDIRECT_BINDING_URI);
```

Comment:
yes, this is a bit unclear functionality as UAA/Spring saml IDP do not support redirect binding as defined by specification (http://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf  424). But, all other SAML IDPs i've seen do support it. OTOH support for redirect binding is coming to UAA too (https://github.com/cloudfoundry/uaa/pull/656). Maybe preferred binding method should be configurable?